dali/python/nvidia/dali/_debug_mode.py

```diff
@@ -0,0 +1,332 @@
+        self.pipeline._debug_on = False
+        self.pipeline._external_source_debug = False
+        if not self.pipeline._subpipelines_built:
+            self.pipeline._subpipelines_built = True
+        return super(type(self.pipeline), self.pipeline).__exit__(*exc)
+
+
+class _PipelineDebug(_pipeline.Pipeline):
+    """Debug mode for pipeline. Allows access to data inside the pipeline execution by wrapping all
+     operators inside their pipelines"""
+
+    def __init__(self, exec_func, **kwargs):
+        kwargs['debug'] = False
+        super().__init__(**kwargs)
+        kwargs['exec_pipelined'] = False
+        kwargs['exec_async'] = False
+        self._debug_on = False
+        self._external_source_debug = False
+        self._es_input_name = 'input_'
+        self._es_kwarg_name = 'kwarg_'
+        self._subpipeline_kwargs = kwargs
+        self._subpipelines = {}
+        self._external_source_pipelines = {}
+        self._feed_input_data = {}
+        self._subpipelines_built = False
+        self._cur_subpipeline_id = -1
+        self._exec_func = exec_func
+        self._runner = _PipelineDebugRunner(self)
+
+    def __enter__(self):
+        raise RuntimeError("Currently pipeline in debug mode works only with `pipeline_def` decorator."
+                           "Using `with` statement is not supported.")
+
+    def build(self):
+        """Build the pipeline.
+        
+        Refer to :meth:`Pipeline.build() <nvidia.dali.Pipeline.build>` for details."""
+        self._built = True
+
+    def run(self):
+        """Run the pipeline and return the result."""
+        import numpy as np
+        if not self._built:
+            raise RuntimeError('Pipeline must be built first.')
+        with self._runner:
+            res = self._exec_func()
+            if res is None:
+                res = ()
+            elif not isinstance(res, tuple):
+                res = (res,)
+
+            # Transforming all variables to TensorLists.
+            return tuple([val.get() if isinstance(val, DataNodeDebug) else _Tensors.TensorListCPU(
+                np.tile(val, (self._max_batch_size, *[1]*np.array(val).ndim))) for val in res])
+
+    def feed_input(self, data_node, data, **kwargs):
+        """Pass data to an ExternalSource operator inside the pipeline.
+        
+        Refer to :meth:`Pipeline.feed_input() <nvidia.dali.Pipeline.feed_input>` for details."""
+        if not self._built:
+            raise RuntimeError("Pipeline must be built first.")
+
+        if data_node not in self._external_source_pipelines:
+            # Saving data, because pipeline hasn't been run yet.
+            if data_node not in self._feed_input_data:
+                self._feed_input_data[data_node] = []
+            self._feed_input_data[data_node].append((data, kwargs))
+        else:
+            self._external_source_pipelines[data_node].feed_input(data_node, data, **kwargs)
+
+    @staticmethod
+    def _classify_data(data):
```

Comment:
Probably we should slap a comment here from that google doc, where we listed how we recognize data for external source vs constants.
Tl;dr - if something can be recognized as batch without false negatives when it has to be constant, we use external source, so list of non-primitive types and explicit batch types (TensorList)